### PR TITLE
RTC retry improvements

### PR DESCRIPTION
**Pull request by keianhzo**
- Added some more missing retry cases
- Added missing Recv watchdog stop when disconnecting
- Fixed comment typos
- Reduced retry time to 2secs
- RTC Log clear/download
- Pause remote server stats when collapse (start with the remote panel collapsed)

### CODE REVIEW

File: src/naf-dialog-adapter.js [hunk 1]
```diff
@@ -143,26 +192,159 @@ export default class DialogAdapter {
...
+   * Checks the Send Transport ICE status and restarts it in case is in failed or disconnected states.
+   * This is called by the Send Transport "connectionstatechange" event listener.
+   * @param {boolean} connectionState The transport connnection state (ICE connection state)
+   */
+  checkSendIceStatus(connectionState) {
+    // If the ICE connection state failed we force an ICE negotiation
+    if (connectionState === "failed" || connectionState === "disconnected") {
+      this.restartSendICE(false);
+    } else if (connectionState === "connected") {
+      this._iceReconnectSendTaskId && clearTimeout(this._iceReconnectSendTaskId);
```
Comment (robinkwilson) [inline src/naf-dialog-adapter.js:246]: Is this a new javascript syntax for a simple if statement? That's awesome!

File: src/naf-dialog-adapter.js [hunk 2]
```diff
@@ -143,26 +192,159 @@ export default class DialogAdapter {
...
+  }
+
   /**
    * Restart ICE in the underlying receive peerconnection.
+   * @param {boolean} force Forces the execution of the reconnect.
    */
-  async restartRecvICE() {
+  async restartRecvICE(force = true) {
+    // Renegotiate ICE in case ICE state is failed, disconnected. The last condition cover the cases where the
+    // Send transport has been create but for whatever reason is stuck in new state and doesn't gather candidates.
```
Comment (robinkwilson) [inline src/naf-dialog-adapter.js:283]:
```suggestion
    // send transport has been create but for whatever reason is stuck in new state and doesn't gather candidates.
```
semantic change, let's me know it's in the middle of the sentence, up to you to change.

File: src/naf-dialog-adapter.js [hunk 3]
```diff
@@ -143,26 +192,159 @@ export default class DialogAdapter {
...
     if (!this._recvTransport?._closed) {
-      this.emitRTCEvent("log", "RTC", () => `Restarting receive transport ICE`);
-      this._recvTransport.restartIce({ iceParameters: this._recvIceParameters });
+      if (
+        force ||
+        this._recvTransport.connectionState === "failed" ||
+        (!isFirefox && this._sendTransport.connectionState === "disconnected") ||
+        (this._recvTransport.connectionState === "new" && this._lastRecvConnectionState === "failed")
```
Comment (robinkwilson) [inline src/naf-dialog-adapter.js:289]: Should this block be its own function? It _seems_ like it's the same code twice here and line 214

File: src/naf-dialog-adapter.js [hunk 4]
```diff
@@ -143,26 +192,159 @@ export default class DialogAdapter {
...
-      this._sendTransport.restartIce({ iceParameters: this._sendIceParameters });
+      // Renegotiate ICE in case ICE state is failed, disconnected. The last two conditions cover the cases where the
+      // Send transport has been create but for whatever reason is stuck in new state and doesn't gather candidates.
+      if (
+        force ||
+        this._sendTransport.connectionState === "failed" ||
+        (!isFirefox && this._sendTransport.connectionState === "disconnected") ||
+        (this._sendTransport.connectionState === "new" && this._lastSendConnectionState === "failed")
+      ) {
+        this.emitRTCEvent("log", "RTC", () => `Restarting send transport ICE`);
```
Comment (robinkwilson) [inline src/naf-dialog-adapter.js:220]: See line 289 (since looks like this is the same codeblock)

**Review by robinkwilson — COMMENTED** [review]
A few comments + small code refactor. I appreciate your fantastic comments that make the functions/code blocks really clear.

File: src/naf-dialog-adapter.js [hunk 5]
```diff
@@ -143,26 +192,159 @@ export default class DialogAdapter {
     }
   }
 
+  /**
+   * This applies to the ICE restart method below.
+   * Why do we need new ICE servers? In case the ICE fails we need to re-negotiate new candidates
+   * but most probably our TURN credentials have expired so we need to request new credentials and
+   * update the TURN servers to be able to gather new candidates.
+   * Firefox doesn't support hot updating ICE servers through mediasoup's updateIceServers (which
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/naf-dialog-adapter.js:200]: Mediasoup is the underlying WebRTC stack that we are using in Dialog.

File: src/naf-dialog-adapter.js [hunk 6]
```diff
@@ -143,26 +192,159 @@ export default class DialogAdapter {
     }
   }
 
+  /**
+   * This applies to the ICE restart method below.
+   * Why do we need new ICE servers? In case the ICE fails we need to re-negotiate new candidates
+   * but most probably our TURN credentials have expired so we need to request new credentials and
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/naf-dialog-adapter.js:198]: Fixed

File: src/naf-dialog-adapter.js [hunk 7]
```diff
@@ -143,26 +192,159 @@ export default class DialogAdapter {
...
+
   /**
    * Restart ICE in the underlying send peerconnection.
    */
-  async restartSendICE() {
+  async restartSendICE(force = true) {
     if (!this._sendTransport?._closed) {
-      this.emitRTCEvent("log", "RTC", () => `Restarting send transport ICE`);
-      this._sendTransport.restartIce({ iceParameters: this._sendIceParameters });
+      // Renegotiate ICE in case ICE state is failed, disconnected. The last two conditions cover the cases where the
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/naf-dialog-adapter.js:212]: Updated

File: src/naf-dialog-adapter.js [hunk 8]
```diff
@@ -143,26 +192,159 @@ export default class DialogAdapter {
...
+      // Send transport has been create but for whatever reason is stuck in new state and doesn't gather candidates.
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/naf-dialog-adapter.js:213]: Fixed

File: src/naf-dialog-adapter.js (hunk 1 above)
Comment (keianhzo) [reply src/naf-dialog-adapter.js:246]: You can see it that way, it's just a slightly more concise and idiomatic way of doing a condition check and save some lines of code. Same as the safe call operator (?.)

File: src/naf-dialog-adapter.js (hunk 2 above)
Comment (keianhzo) [reply src/naf-dialog-adapter.js:283]: Updated

File: src/naf-dialog-adapter.js (hunk 3 above)
Comment (keianhzo) [reply src/naf-dialog-adapter.js:289]: As they are using different transports that function would end up not saving that much code lines. But I just realized that I was referencing the wrong transport in `restartRecvICE`. I've fixed that.

Comment (robinkwilson) [reply src/naf-dialog-adapter.js:291]: Nice catch!

**Review by robinkwilson — APPROVED** [review]
LGTM

**Comment by vincentfretin** [comment]
Be careful about the RTCPeerConnection disconnected state. Did you try on iOS? Doing a ICE restart right away when a RCTPeerConnection goes in disconnected state is a bad idea. You should at least wait a few seconds to see if the state doesn't go back to connected by itself. Did you see my recent tests on the janus adapter on https://github.com/mozilla/naf-janus-adapter/pull/107#issuecomment-744591501 ?

**Comment by keianhzo** [comment]
@vincentfretin Thanks for that. For what I have seen the recovery time from disconnects can vary and some times they don't even recover at all and transition to a failed state depending of the type of disconnection. Anyway I've introduced a delay of two secs for ICE starts to give a chance for recovery.

**Comment by vincentfretin** [comment]
Great work @keianhzo You'll see if 2s is enough. Did you test the changes on Safari iOS?

In my app I'm using a modified version of naf-janus-adapter 3.0.20 and finally I don't do the reconnect if a peer connection is in state disconnected, I reverted that. It does a full reconnect after 10s if a peer connection is in state failed or after 40s if the websocket timed out. I could have tried to reconnect if a peer connection is in state disconnected for 10s, but I didn't do it because the janus session is only ended after 40s of inactivity anyway. See the discussion https://github.com/mozilla/janus-plugin-sfu/pull/73

I don't know how it works with mediasoup, does mediasoup detects that the connection is dead and ends it? Can you have several send transport for the same peerId? With janus sfu, we can only have one publisher for a user id, this is why I can't reconnect before 40s.

**Comment by keianhzo** [comment]
The `RTCPeerConnection` state transition does not depend on Mediasoup but on the underlying browser WebRTC implementation and that varies across browsers. From my experience most browsers recover from a disconnect but there are certain cases where they don't or the don't do it fast enough. I could emulate one of this cases in Chrome quickly bringing down and up the TURN server while using TURN. In that case FF seems to recover quickly but Chrome stays in disconnected state. That's why the ICE restart in that case. Even in the case where Chrome recovers from that at some point, we rather not letting the user down for so long.

**Pull request merged**